ConfigMap as a ValueSource in Param in TaskRuns and PipelineRuns

mostafaCamel opened the pull request:
This proposal is to be able to reference `ConfigMap` as a value source for a `Param` in `TaskRun` or `PipelineRun` . This is to support Kubernetes native options (ConfigMap) as value source along with direct value passed to `TaskRun` and `PipelineRun`.

Note: this proposal is basically picking up upon the closed (unmerged) [proposal](https://github.com/tektoncd/community/pull/868).

mostafaCamel commented:
For whatever it's worth, I've made the code change in my fork (I have yet to add documentation and tests)
https://github.com/mostafaCamel/pipeline/commit/0a3f5169e6c0f21a5a72dc3e6fc1a5b7f6dd7aa7

waveywaves commented:
added the TEP as an agenda item to be discussed in the upcoming meeting

mostafaCamel commented:
Thanks Vibhav! I am new to the project so let me know if I need to/can attend

waveywaves commented:
@mostafaCamel you are welcome to join the working group calls ! Check the github.com/tektoncd/community repo for more info 🙂

waveywaves commented on teps/0161-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,305 @@
+### Use Cases
+
+- Task and Pipeline authors will be able to easily use values from ConfigMaps in properties that do not support a script (example: `steps[*].image` in Task)
+- Catalog users will be able to use configmaps to populate in their PipelineRun/TaskRun values required by the Pipeline/Task
```
should we add support for StepActions as well ?

waveywaves replied:
This question was resolved in the last meeting. The answer is that this change applies to runtime objects like PipelineRuns and TaskRuns and not configuration objects like StepActions etc.

vdemeester commented:
/assign

waveywaves commented:
/assign

mostafaCamel commented on teps/0161-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,305 @@
+- For each of the params, if the param has a non-nil `valueFrom` field:
+    - look for the configmap with name `Param.ValueFrom.ConfigMapKeyRef.Name` inside the same k8s namespace of the PipelineRun/TaskRun. PipelineRun to fail if the configmap is not found
+    - inside the configmap, look for the key `Param.ValueFrom.ConfigMapKeyRef.Key`. PipelineRun to fail if the key is not found and if `Param.ValueFrom.ConfigMapKeyRef.Optional` is false (this is the default value of this bool)
+    - Marshal the value obtained from the configmap into a byte array then convert it into a `ParamValue` object via the `*ParamValue.UnmarshalJSON` custom method
```
During the working group meeting today, it was suggested we can support inline parameters by using the `Type` field. I checked and this field is only in ParamSpec and ParamValue , not in Param
https://github.com/tektoncd/pipeline/blob/990917dffc997071df470be66aed30a10338a5e0/pkg/apis/pipeline/v1/param_types.go#L191-L196

What I can do is to add an optional `Type` field inside the ValueSource struct. Then:
- during `createParamSpecFromParam` in validation (for inline parameters) , the ParamSpec type will be populated based on the ValueSource type (and will default to string if ValueSource.Type is not defined by the user).
- during the value fetching from the configmap and the creation of a new `ParamValue` instance, the Type in this ParamValue will be validated against (i.e. fail if different than) the Type (if defined) in ValueSource

SaschaSchwarze0 commented:
fyi ... when we implemented this in Shipwright, we used a different approach to access the value eliminating the need to load it but delegating this to Kubernetes. That way, we are also doing this for Secrets, which through the way it is proposed here, certainly is not possible. On the other hand, our approach limits the usage of such parameters to only cmd, args and env.

https://github.com/shipwright-io/community/blob/main/ships/0025-strategy-parameters-enhancements.md

mostafaCamel commented:
> fyi ... when we implemented this in Shipwright, we used a different approach to access the value eliminating the need to load it but delegating this to Kubernetes. That way, we are also doing this for Secrets, which through the way it is proposed here, certainly is not possible. On the other hand, our approach limits the usage of such parameters to only cmd, args and env.
> 
> https://github.com/shipwright-io/community/blob/main/ships/0025-strategy-parameters-enhancements.md

My understanding Tekton tasks already support value sources (config maps and secrets) for cmd, args and env at the container/step level
https://github.com/tektoncd/pipeline/blob/47bcb0937f2882879c1a5e11492203fe618b730c/docs/variables.md?plain=1#L92

SaschaSchwarze0 commented:
> > fyi ... when we implemented this in Shipwright, we used a different approach to access the value eliminating the need to load it but delegating this to Kubernetes. That way, we are also doing this for Secrets, which through the way it is proposed here, certainly is not possible. On the other hand, our approach limits the usage of such parameters to only cmd, args and env.
> > https://github.com/shipwright-io/community/blob/main/ships/0025-strategy-parameters-enhancements.md
> 
> My understanding Tekton tasks already support value sources (config maps and secrets) for cmd, args and env at the container/step level https://github.com/tektoncd/pipeline/blob/47bcb0937f2882879c1a5e11492203fe618b730c/docs/variables.md?plain=1#L92

For env, sure. That's Kube machinery. I just said that we in Shipwright used that Kube capability to read those parameters that a Shipwright user defined on BuildRun level to come from a secret or configMap into environment variables on the TaskRun that we create for that BuildRun (with exactly what you point to) and then we reference the environment variable value in the param value using `$(ENV_VAR)` syntax.

Theoretically, Tekton could do the same one level deeper when it replaces the values in the PodSpec. Instead of replacing `$(param.my-param)` with the value from the ConfigMap key, you would replace it with `$(SOME_ENV_VAR_THAT_TEKTON_ADDS)` and in the env section of the container, you load the value of that env var from the ConfigMap. And that also works for secrets because you don't have to put the plain value into the Pod spec.

But is just a hint for another option that you may consider or not.

mostafaCamel commented:
Ah ok, I now understand. Thanks for the explanation!

vdemeester commented on teps/0161-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
````diff
@@ -0,0 +1,305 @@
+    name: hello-goodbye
+  params:
+  - name: username
+    valueFrom:
+          configMapKeyRef:
+            name: game-demo
+            key: KEY_IN_CONFIGMAP
+```
````
nit: the indentation seems off.

vdemeester commented on teps/0161-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,305 @@
+
+- At the start of the first reconcile, the value will be obtained from the configmap (the PipelineRun will fail if the value fetching fails).
+
+- After the first reconcile run, the k8s resource PipelineRun object will be patched with the resolved parameter value so that subsequent reconcile runs do not fetch the value from the configmap.
```
Question on this: does this mean the would remove `valueFrom` and become a `value` field ? We would loose some information if that's the case.

I would rather keep `valueFrom` and have a field called `resolvedValue` (or something) that would contain the value. This way, we do not remove the "user input" that was `valueFrom`. It adds one more field but it's more explicit and more traceable (we know from where the value came from).

cc @tektoncd/core-maintainers

mostafaCamel replied:
Thanks @vdemeester. Correct. This means that we are losing this info.  I thought about having a `resolvedValue` field but did not propose it because:

- There is no getter method for param.Value and the field is accessed directly in tons of occurrences across the codebase. A new field `resolvedValue` would require the new getter method + a ton of refactoring. I don't think it is worth it but I can be wrong. Let me know what you think.

- Some info is already lost in the "sub-objects" of the PipelineRun k8s object. Example: the params that are replaced in the PipelineTasks so I didn't see it as a big issue



Somewhat adjacent to this discussion: if we don't go the getter method route, we can add a private (so that it is not (de)serialized)and cannot be set by the user in the yaml)  boolean field `param.valueFrom.isResolved` which is set to true during the value resolution .... and then during the k8s object patching, this will validate that thie `value` was done by the reconciler and not the user who is applying a patched yaml

vdemeester replied:
> Some info is already lost in the "sub-objects" of the PipelineRun k8s object. Example: the params that are replaced in the PipelineTasks so I didn't see it as a big issue

Hmmm, not in the `spec.…` fields though, those should stay exactly as the user created. If not, I am/would be a bit concerned by that as well…

> There is no getter method for param.Value and the field is accessed directly in tons of occurrences across the codebase. A new field resolvedValue would require the new getter method + a ton of refactoring. I don't think it is worth it but I can be wrong. Let me know what you think.

I am really thinking only in terms of the API and what the user sees, not really taking into account implementation "problems" yet 😝.  As a user, I think I want to be able to "trace" where the param value was coming from, those I don't really want to loose the information on which configmap and which key it was coming from.

vdemeester reviewed (changes requested):
I have one important question around the "patching" the object (and thus removing some user input), otherwise 👍🏼

mostafaCamel commented:
Rebased the branch and added a new commit to address the reviews.
- The reconciler will no longer unset param.Value after the first reconciler run

I added an explanation in the risks/mitigation section

mostafaCamel commented on teps/0162-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,313 @@
+
+## Proposal
+
+- The user will be able to define one or more `Param`s in a PipelineRun or a TaskRun with `spec.params[*].valueFrom.configMapKeyRef.name` and `spec.params[*].valueFrom.configMapKeyRef.key` instead of the usual hardcoded `spec.params[*].value` 
```
tbh I did not think about it but it should be doable

mostafaCamel replied:
Actually, thinking about it again. I am not sure what is the source for the variable substitution in a pipelinerun parameter. The only thing I can think of is TriggerTemplates but TriggerTemplate is deprecated in v1 and the plan (so far) is to implement this new feature in v1 only so there would be no need to support substitution

afrittoli commented on teps/0162-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,313 @@
+- If the check is true:
+   - patch the k8s resource with the new Params array (by calling `c.PipelineClientSet.TektonV1().PipelineRuns(pr.Namespace).Patch`)
+
+#### Relaxing the validation rules for the PipelineRun/TaskRun resource patch/updates
```
I think this is not needed if we cache the value in the status instead

afrittoli reviewed (commented):
Thanks for this, I'm in favour of this feature.

My primary concern is with the proposed plan to update the spec once a `ValueFrom` is resolved to its value. I agree that we should cache the value, but we can do that in the status, as with any other thing that we cache (e.g. spec of remotely resolved tasks or step actions).

mostafaCamel commented on teps/0162-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,313 @@
+  params:
+  - name: username
+    valueFrom:
+          configMapKeyRef:
```
I did not think about it but it should be doable (`namespace` property in configMapKeyRef with a default fallback to the same namespace of the TaskRun/PipelineRun

mostafaCamel commented on teps/0162-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,313 @@
+
+- At the start of the first reconcile, the value will be obtained from the configmap (the PipelineRun will fail if the value fetching fails).
+
+- After the first reconcile run, the k8s resource PipelineRun object will be patched with the resolved parameter value so that subsequent reconcile runs do not fetch the value from the configmap.
```
Correct

mostafaCamel commented on teps/0162-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,313 @@
+
+
+### New API Object's validation
+- Changes to the [implementation](https://github.com/tektoncd/pipeline/blob/5b082b1106753e093593d12152c82e1c4b0f37e5/pkg/apis/pipeline/v1/param_types.go#L493) of the custom methods `*ParamValue.UnmarshalJSON` and `ParamValue.MarshalJSON` to allow empty `ParamValue` fields
```
The pipeline project has [recently adopted](https://github.com/tektoncd/pipeline/blob/bb8b8bb8bf5fcc8ec74977b50990a6180d0d5a62/go.mod#L3) go1.24 which has a new [omitzero feature](https://tip.golang.org/doc/go1.24#encodingjsonpkgencodingjson) that we can add tot he ParamValue field inside the Param struct

So we can skip the changes to the custom marshalling methods (but we will still need to verify in the validation either ParamValue or ValueFrom are passed)

mostafaCamel commented on teps/0162-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,313 @@
+
+### New API Object's validation
+- Changes to the [implementation](https://github.com/tektoncd/pipeline/blob/5b082b1106753e093593d12152c82e1c4b0f37e5/pkg/apis/pipeline/v1/param_types.go#L493) of the custom methods `*ParamValue.UnmarshalJSON` and `ParamValue.MarshalJSON` to allow empty `ParamValue` fields
+    - There is a [roundTripPatch](https://github.com/tektoncd/pipeline/blob/5b082b1106753e093593d12152c82e1c4b0f37e5/vendor/knative.dev/pkg/webhook/resourcesemantics/defaulting/defaulting.go#L324) which occurs before the validation (i.e the yaml is deserialized into an object, then serialized, then desrialized again then the validation occurs). So the change in the implementation of these custom  ...
```
My bad. I meant the `*ParamValue.UnmarshalJSON` and `ParamValue.MarshalJSON` methods in our source code. No need to change anything on Knative.

mostafaCamel commented:
The eventual getter method I would need to implement will be used to replace every invocation similar to this: https://github.com/tektoncd/pipeline/blob/bb8b8bb8bf5fcc8ec74977b50990a6180d0d5a62/pkg/reconciler/pipelinerun/resources/apply.go#L110

mostafaCamel commented on teps/0162-configmap-as-a-valuesource-in-param-in-taskruns-and-pipelineruns.md:
```diff
@@ -0,0 +1,313 @@
+    - The webhook cannot poll for the configmap existence so this check will be done by the reconciler later (in the controller).
+
+### Fetching values from the configmap
+This value resolution will be done one time only (if param.ValueFrom is defined and param.Value is empty).
```
Could another option would be to populate param.Value during the first mutating webhook call (before the initial persistence of PipelineRun) by adding some custom logic in the defaulting (also called mutating)  admission webhook ? We are calling knative's `defaulting.NewAdmissionController` [with some arguments](https://github.com/tektoncd/pipeline/blob/d5048903a9012f505349751bbc2644acfef28477/cmd/webhook/main.go#L109-L128)

vdemeester commented:
@mostafaCamel the PR needs a rebase.
@tektoncd/core-maintainers can you review this ? Overall, I agree with the feature, but we need to figure things out implementation wise. It shouldn't block the TEP though (and we could reduce the implementation detail appearing on it if it does block merging this).
